### PR TITLE
sql: Add more variables for controlling RBAC

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds two additional system/session parameters to turn RBAC on and off. The parameters that exist are as follows:

  - A system parameter called "enable_ld_rbac_checks". This parameter will only be able to be set by Launch Darkly, no users can toggle this parameter. When this is off, then no user can enable RBAC. When this is on, then a user can enable RBAC through one of the other parameters. This acts as a global off switch for RBAC in case we ever need to ensure that RBAC is off for a customer.
  - A system parameter called "enable_server_rbac_checks". This parameter can be set by any superuser. When this is on, then RBAC is enabled for all users, as long as "enable_ld_rbac_checks" is also on.
  - A session parameter called "enable_session_rbac_checks". This parameter is session local and can be set by any user. When this is on, then RBAC is enabled for the current session, as long as "enable_ld_rbac_checks" is also on. The intention is that this can be used to test RBAC in a single session, without having to turn it on for everyone.

Part of #11579

### Motivation
* This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by jkosh44** [comment]
@chaas, @ggnall FYI here are all the parameters for RBAC. The commit message explains all of them in depth.

**Review by parker-timmerman — APPROVED** [review]
Looks good to me! Just some small changes w.r.t. code layout

File: misc/python/materialize/checks/mzcompose_actions.py [hunk 1]
```diff
@@ -85,7 +85,8 @@ def execute(self, e: Executor) -> None:
         )
 
         if self.base_version >= MzVersion(0, 46, 0):
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply misc/python/materialize/checks/mzcompose_actions.py:87]: @def- What's the process/logic for incrementing this? I'm not really sure what it is so I'd be a bit hesitant to change it.

File: src/adapter/src/notice.rs [hunk 2]
```diff
@@ -104,7 +105,8 @@ impl AdapterNotice {
             AdapterNotice::DroppedActiveDatabase { name: _ } => Some("Choose a new active database by executing SET DATABASE = <name>.".into()),
             AdapterNotice::DroppedActiveCluster { name: _ } => Some("Choose a new active cluster by executing SET CLUSTER = <name>.".into()),
             AdapterNotice::ClusterReplicaStatusChanged { status, .. } if *status == ClusterStatus::NotReady => Some("The cluster replica may be restarting or going offline.".into()),
-            AdapterNotice::RbacDisabled => Some("To enable RBAC run `ALTER SYSTEM SET enable_rbac_checks TO true` as a superuser.".into()),
+            AdapterNotice::RbacSystemDisabled => Some("To enable RBAC please reach out to support with a request to turn RBAC on.".into()),
+            AdapterNotice::RbacUserDisabled => Some("To enable RBAC globally run `ALTER SYSTEM SET enable_server_rbac_checks TO true` as a superuser. TO enable RBAC for just this session run `SET enable_session_rbac_checks TO TRUE`.".into()),
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/notice.rs]: Updated

File: src/adapter/src/command.rs [hunk 3]
```diff
@@ -114,7 +114,24 @@ pub enum Command {
...
+            | Command::DumpCatalog { session, .. }
+            | Command::CopyRows { session, .. }
+            | Command::GetSystemVars { session, .. }
+            | Command::SetSystemVars { session, .. }
+            | Command::Terminate { session, .. } => Some(session),
+            Command::CancelRequest { .. } => None,
+        }
+    }
+
+    pub fn session_mut(&mut self) -> Option<&mut Session> {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/command.rs:134]: Good catch. I made that change to `add_notice`, but `apply_external_metadata_updates` also needs `session_mut`.

**Comment by jkosh44** [comment]
@chaas @ggnall Before merging, I just want to double check that we all agree that these three params are the right design. Are you two on board?

File: src/adapter/src/coord/sequencer.rs [hunk 4]
```diff
@@ -469,4 +467,14 @@ impl Coordinator {
         self.transient_id_counter += 1;
         Ok(GlobalId::Transient(id))
     }
+
+    fn maybe_send_rbac_notice(&self, session: &mut Session) {
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer.rs:471]:
```suggestion
    fn maybe_send_rbac_notice(&self, session: &Session) {
```

File: src/sql/src/func.rs [hunk 5]
```diff
@@ -3028,6 +3028,9 @@ pub static MZ_INTERNAL_BUILTINS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(
     use ParamType::*;
     use ScalarBaseType::*;
     builtins! {
+        "is_rbac_enabled" => Scalar {
```
Comment (mjibson) [inline src/sql/src/func.rs:3031]: This needs an entry in funcs.yml.

**Review by mjibson — APPROVED** [review]

File: src/sql/src/func.rs (hunk 5 above)
Comment (jkosh44) [reply src/sql/src/func.rs:3031]: Done

**Comment by aalexandrov** [comment]
@jkosh44 can you share more about the rollout plan and the intended lifespan of all those features? Are we going to phase them out eventually or are these things that we'll stick around forever?

**Comment by jkosh44** [comment]
> @jkosh44 can you share more about the rollout plan and the intended lifespan of all those parameters? Are we going to phase them out eventually or are we going to keep them forever?

I'll update the design doc (https://github.com/MaterializeInc/materialize/pull/17702) to include these parameters. The goal is eventually to phase out all parameters except for `enable_server_rbac_checks`. `enable_server_rbac_checks` will default to `true` for all environments, but if a customer doesn't care about RBAC, then they can disable it for their environment.

**Comment by jkosh44** [comment]
> If `enable_server_rbac_checks` is the only one that will remain eventually maybe we can bite the bullet short-term and give it a less specific name (`enable_rbac_checks`), so that people are not confused about the `server` part once the other two are gone.
> 
> It's probably more of a @ggnall call here, just highlighting that one of the three variable names is something that will stick forever.

I think that's a really good point. I agree that we should rename it now to `enable_rbac_checks`.

**Comment by ggnall** [comment]
Late but agree!
>I agree that we should rename it now to enable_rbac_checks.

**Comment by jkosh44** [comment]
I just updated the name.

File: misc/python/materialize/checks/mzcompose_actions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (def-) [reply misc/python/materialize/checks/mzcompose_actions.py]: The idea is that the test has to keep working even if it is started with an older Mz version, which is later upgraded. We use the initial version here to limit the features.

**Pull request merged**